Login screen

lmillet3804 opened the pull request:
Added a skeleton for the login screen...certainly not pretty nor finished, but at least as some basic fields so far.

Also has forgot password and create new account pages

lmillet3804 committed:
Expo app-safe changes

-Changed spans and ps to Texts
-Changed divs to Views

lmillet3804 committed:
Skeleton of Login flow

-added components to main login screen
-added forgot password Page
-added create Page

TheGoldenPatrik1 commented on app/auth/create.js:
Can we rename this to `createAccount`? And likewise can we rename `forgot` to `forgotPassword`? I think it would be beneficial to have more clear naming.

lmillet3804 replied:
I agree, I thought I'd want to keep convention of hyphenating words for urls so I left it as one word, but come to think of it...we never really see the url in an app, so it wouldn't matter

TheGoldenPatrik1 commented on app/auth/create.js:
```diff
@@ -0,0 +1,101 @@
+    const [passwordsMatch, setPasswordsMatch] = useState(true)
+
+    useEffect(() => {
+        const re = /^[\w-\\.]+@crimson.ua.edu$/
```
I think we could also support `@ua.edu` emails. If you agree, then also change the TextInput label to say "UA Email" rather than "Crimson Email."

lmillet3804 replied:
@ua.edu emails are for professors, should they be able to log in? Do they get meal plans?

TheGoldenPatrik1 replied:
They could certainly buy a meal using the app. Also, I got a @ua.edu email for my research job, so it's not just professors. Perhaps this is an @ardymarky question though.

TheGoldenPatrik1 replied:
Oh also, let's avoid the lazy variable name `re` and say something like `emailRegex`. :)

TheGoldenPatrik1 commented on app/auth/create.js:
```diff
@@ -0,0 +1,101 @@
+    }, [email])
+
+    useEffect(() => {
+        if (passwordConfirm.length > 0 && passwordConfirm == password) {
```
The abstract equality operator (`==`) here shouldn't be an issue but let's change this to the strict equality operator (`===`) since that's what you should almost always use. For more info: https://stackoverflow.com/questions/359494/which-equals-operator-vs-should-be-used-in-javascript-comparisons

TheGoldenPatrik1 commented on app/auth/create.js:
```diff
@@ -0,0 +1,101 @@
+        await dispatch(createUser)
+    }
+
+    const isLoggedIn = useSelector(selectIsLoggedIn)
```
This line and the `useEffect` hook that follow it should be automatically handled by the `AuthPage` component.

TheGoldenPatrik1 commented on app/auth/login.js:
```diff
@@ -24,10 +27,35 @@ const Login = () => {
+            </Text>
+            <Text
+                onPress={() => router.replace('/auth/create')}
+                style={{ color: 'blue', textDecorationLine: 'underline' }}
```
This duplicate style could be moved into a StyleSheet. Check the `Container` component for an example of that.

TheGoldenPatrik1 replied:
Per https://reactnative.dev/docs/stylesheet let's keep the StyleSheet's outside of the main render function. Again, see the Container component.

TheGoldenPatrik1 commented on app/auth/create.js:
```diff
@@ -0,0 +1,101 @@
+        } else {
+            setPasswordsMatch(false)
+        }
+    }, [isLoggedIn])
```
Why is the dependency for this useEffect hook the isLoggedIn variable? That's not making sense to me. Shouldn't the dependencies be passwordConfirm and password, so that we perform a check when either of them changes?

TheGoldenPatrik1 replied:
Shouldn't `password` also be a dependency? Otherwise someone could bypass the check by getting the `confirmPassword` equal to `password` and then changing `password` after the fact.

TheGoldenPatrik1 reviewed (changes requested):
Great work! Glad to see you already figuring out react-native, redux, react-native-paper, etc!

lmillet3804 committed:
addressed PR comments:

-rename create.js and forgot.js
-removed isLoggedIn useEffect hook on createAccount and login
-consolidated styles on login

lmillet3804 committed:
addressed more PR comments:

-moved styles out of render function
-added password as a dependency to the useEffect hook
-added eqeqeq rule

samuelmgaines commented on app/auth/createAccount.js:
We need to work together tomorrow to determine our password requirements. Right now, I'm requiring 8 characters. It might be smart to require more. Whatever we decide, you should probably enforce it on the front end along with me enforcing it on the back end. Don't worry about this now, but we will probably need to do this.
